client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedPusher.java

```diff
@@ -219,6 +230,39 @@ public void insertRecord(
       pageCursor += recordSize;
     }
     inMemSorter.insertRecord(recordAddress, partitionId);
+
+    return true;
+  }
+
+  public void triggerPush() throws IOException {
+    dataPusher.checkException();
+    if (pushing) {
+      throw new IOException("Re-trigger push!");
+    }
+    pushing = true;
+    Thread thread =
+        new Thread(
+            () -> {
+              try {
+                pushData();
+                pushing = false;
+              } catch (IOException ie) {
+                dataPusher.setException(ie);
+              }
+            });
+    thread.start();
+  }
+
+  public void waitPushFinish() throws IOException {
+    dataPusher.checkException();
+    while (pushing) {
+      try {
+        Thread.sleep(100);
```

Comment:
I think waiting is not a good solution here. Maybe lock or sync will be better here.

Reply:
This waiting might waste time.

Reply:
A thread is pushing data while another is querying its state. Change pushing to lock could be enough to avoid waiting.